### PR TITLE
Add ajaxUILoc XSS protection

### PR DESCRIPTION
**Pull request by daniel-samson**
## Description

This change attempts to fix a possible XSS issue with ajaxUILoc in the url, by validating the url.

it assumes that you do not need brackets "()", as these can be used to call JavaScript functions / objects

## Motivation and Context

Removes potential XSS attacks using ajaxUILoc.

## Recommendation
I would recommend that we remove the ajaxUI from a future version of SuiteCRM.

## How To Test This

- Rebuild minified js
- Rebuild js groupings
- Clear Browser Cache
- open URL:
http(s)://path/to/instance/index.php?action=ajaxui&LicState=check#ajaxUILoc=javascript:alert(1)

Failure:
if browser open JS alert which show number 1.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [ ] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php/Contributing_to_SuiteCRM) guidelines.

### CODE REVIEW

File: jssource/src_files/include/javascript/ajaxUI.js [hunk 1]
```diff
@@ -168,14 +168,26 @@ SUGAR.ajaxUI = {
...
         if(YAHOO.lang.trim(url) != "")
         {
+
+          // ajaxUILoc XSS protection:
+          // window.location = url; is vulnerable to XSS attack
+          // Check for valid url
+          // Expects url encoded versions of index.php?module=Home&action=index&parentTab=All
+          var testUrl = decodeURIComponent(url);
+          if (
+            /^index.php([A-Z]|[%=+#&\?\[\]\.]|[a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+/i.test(testUrl) === false
```
Comment (gymad) [inline jssource/src_files/include/javascript/ajaxUI.js]: A little bit more specific and strict regex:
`/^index.php?(([A-Z]|[a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+[%=+#&\[\]\.]*)+/i.test(testUrl)`

File: jssource/src_files/include/javascript/ajaxUI.js [hunk 2]
```diff
@@ -186,8 +198,21 @@ SUGAR.ajaxUI = {
...
                 return;
+            }
+
+          // ajaxUILoc XSS protection:
+          // window.location = url; is vulnerable to XSS attack
+          // Check for valid url
+          // Expects url encoded versions of index.php?module=Home&action=index&parentTab=All
+          var testUrl = decodeURIComponent(url);
+          if (
+            /^index.php([A-Z]|[%=+#&\?\[\]\.]|[a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+/i.test(testUrl) === false
```
Comment (gymad) [inline jssource/src_files/include/javascript/ajaxUI.js]: See regex above. Also can we add a function for this, instead code duplication.

File: modules/ModuleBuilder/javascript/ModuleBuilder.js [hunk 3]
```diff
@@ -79,8 +79,20 @@ if (typeof(ModuleBuilder) == 'undefined') {
...
-                window.location = "index.php?action=ajaxui#ajaxUILoc=" + ajaxLoc;
-                return;
+			  var url = "index.php?action=ajaxui#ajaxUILoc=" + ajaxLoc;
+              // ajaxUILoc XSS protection:
+              // window.location = url; is vulnerable to XSS attack
+              // Check for valid url
+              // Expects url encoded versions of index.php?module=Home&action=index&parentTab=All
+              var testUrl = decodeURIComponent(url);
+              if (
+                /^index.php([A-Z]|[%=+#&\?\[\]\.]|[a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+/i.test(testUrl) === false
```
Comment (gymad) [inline modules/ModuleBuilder/javascript/ModuleBuilder.js]: see above

File: include/javascript/ajaxUI.js [hunk 4]
```diff
@@ -1,6 +1,9 @@
 /*********************************************************************************
  * SugarCRM Community Edition is a customer relationship management program developed by
  * SugarCRM, Inc. Copyright (C) 2004-2013 SugarCRM Inc.
+ 
+ * SuiteCRM is an extension to SugarCRM Community Edition developed by Salesagility Ltd.
+ * Copyright (C) 2011 - 2014 Salesagility Ltd.
```
Comment (mattlorimer) [inline include/javascript/ajaxUI.js]: wrong license

**Pull request closed without merging**